# **General Objectives**

---

The general objective of this project is to analyze data from a [Kaggle Dataset](https://www.kaggle.com/datasets/arsalanrehman/movies-dataset-from-piracy-website) that has been gathered from a pirated website that has a user base of around 2M visitors per month. This data contains more than 20,000+ movies from all industries such as Hollywood, Bollywood, Anime, etc. 

The goal is to describe the data the best as possible, utilizing data science methods, statistical aproaches to the data and the creation of a very detailed report.

As this project will be large, to avoid a single notebook with 10K+ lines of code and markdown cells and a very busy, "filled to the brim with information" notebook, we will work in four diferent jupyter notebook reports, each one holding a step in our EDA.

The steps for the creation of the reports are as follows:

**Notebook 1:**
- Collect and Understanding of the Data.

- Data Prep and Transformation.

**Notebook 2:**
- Univariate Analysis.

**Notebook 3:**
- Multivariate Analysis.

**Notebook 4:**
- Questions, Insights and Answers.

# **Importing Packages and Collecting the Dataset**

---

In [78]:
# Importing libraries needed for the first step of the analysis:
import pandas as pd
import numpy as np
import warnings
import re

# Filtering out warnings:
warnings.filterwarnings('ignore')

# Setting pandas dataframe visualization parameters:
pd.set_option('display.max_columns', 100)

print('Packages collected!')

Packages collected!


In [79]:
# Creating the Dataframe with original data:
data = pd.read_csv('data/movies_dataset.csv', sep=',')

- First rows:

In [80]:
data.head()

Unnamed: 0  IMDb-rating appropriate_for      director downloads      id  \
0           0          4.8               R     John Swab       304  372092   
1           1          6.4           TV-PG   Paul Ziller        73  372091   
2           2          5.2               R  Ben Wheatley     1,427  343381   
3           3          8.1             NaN  Venky Atluri     1,549  372090   
4           4          4.6             NaN  Shaji Kailas       657  372089   

              industry       language   posted_date release_date  run_time  \
0  Hollywood / English        English  20 Feb, 2023  Jan 28 2023       105   
1  Hollywood / English        English  20 Feb, 2023  Feb 05 2023        84   
2  Hollywood / English  English,Hindi  20 Apr, 2021  Jun 18 2021  1h 47min   
3            Tollywood          Hindi  20 Feb, 2023  Feb 17 2023       139   
4            Tollywood          Hindi  20 Feb, 2023  Jan 26 2023       122   

                                           storyline  \
0  Doc\r\n facilitates a fragile truce between th...   
1  Caterer\r\n Goldy Berry reunites with detectiv...   
2  As the world searches for a cure to a disastro...   
3  The life of a young man and his struggles agai...   
4  A man named Kalidas gets stranded due to the p...   

                                        title   views                  writer  
0                                Little Dixie   2,794               John Swab  
1  Grilling Season: A Curious Caterer Mystery   1,002  John Christian Plummer  
2                                In the Earth  14,419            Ben Wheatley  
3                                      Vaathi   4,878            Venky Atluri  
4                                       Alone   2,438        Rajesh Jayaraman

- Sample rows:

In [81]:
data.sample(5)

Unnamed: 0  IMDb-rating appropriate_for                       director  \
4986         4986          5.1             NaN                            NaN   
9457         9457          7.1               R               Elegance Bratton   
9855         9855          8.8             NaN                Xavier Manrique   
5603         5603          5.5              PG  Lasse Hallström, Joe Johnston   
12903       12903          5.1             NaN                   Nick Willing   

      downloads      id             industry        language   posted_date  \
4986        901  331119         Anime / Kids          French  18 Jul, 2020   
9457        468  371991  Hollywood / English         English  17 Feb, 2023   
9855         75  371744  Hollywood / English         English  13 Feb, 2023   
5603     21,899  281292  Hollywood / English  English,French  03 Nov, 2018   
12903     5,058  121188  Hollywood / English         English  19 Feb, 2015   

      release_date  run_time  \
4986   Feb 05 2020  1h 17min   
9457   Dec 02 2022        95   
9855   Feb 03 2023       101   
5603   Nov 02 2018        99   
12903  Dec 27 2014    89 min   

                                               storyline  \
4986   SamSam appears to have it all: his own flying ...   
9457   A\r\n young, gay Black man, rejected by his mo...   
9855   Follows\r\n a New York City family hiding out ...   
5603   A young girl is transported into a magical wor...   
12903  The Hamilton family move into a large country ...   

                                    title   views  \
4986                               SamSam   7,386   
9457                       The Inspection   5,828   
9855                 Who Invited Charlie?   1,905   
5603   The Nutcracker and the Four Realms  91,660   
12903                               Altar  19,640   

                                                writer  
4986                                               NaN  
9457                                  Elegance Bratton  
9855                                   Nicholas Schutt  
5603   Ashleigh Powell, E.T.A. Hoffmann, Marius Petipa  
12903                                     Nick Willing

# **Understanding the Dataset**

---

Before we can start describing and analyzing the Data, it's important to comprehend what variables are present in the Dataset and the values they hold. For that, we will use a simple markdown table explaining the contents of the Dataset:

**Table 1.1: Variable Dictionary and Analysis**

| Variable Name             | Variable Contents                 | Variable Importance for Analysis             | Comments About Variable                                                                        |
|---------------------------|-----------------------------------|----------------------------------------------|------------------------------------------------------------------------------------------------|
| **`Unnamed:0`**           | Row id native to the Dataset      | 🔴 Irrelevant                                 | Not needed since Pandas Dataframes already comes with a Id column                              |
| **`IMDB-rating`**         | Rating of the movie on IMDB       | 🟢 High                                       | None                                                                                           |
| **`appropriate_for`**     | Movie classification rating       | 🟡 Medium                                     | Interesting possibilities for analysis. Be careful with the amount of different rating systems |
| **`director`**            | Name of movie director            | 🟡 Medium                                     | None                                                                                           |
| **`dowloads`**            | Number of dowloads per movie      | 🟢 High                                       | None                                                                                           |
| **`id`**                  | Unique Id per movie               | 🔴 Irrelevant                                 | Same motive as the first variable                                                              |
| **`industry`**            | Industry that produced the movie  | 🟢 High                                       | None                                                                                           |
| **`language`**            | Available languages for the movie | 🟠 Low                                         | Not much important for the analysis as a hole, but not totally irrelevant                      |
| **`posted_date`**         | When the movie was posted on the platform | 🟢 High                                | Very important metric for the analysis                                                         |
| **`released_date`**       | When the movie was released worldwide | 🟢 High                                    | In conjunction with the variable above, opens up lots of analytical possibilities             |
| **`run_time`**            | Runtime of the movie (minutes)    | 🟡 Medium                                      | None                                                                                           |
| **`storyline`**           | Movie synopsis                    | 🔴 Irrelevant                                  | For the pourposes of this analysis, the movie storyline is not needed                          |
| **`title`**               | Movie title                       | 🟢 High                                      | Without a name, there is no movie!                                                             |
| **`views`**               | Number of clicks per movie        | 🟢 High                                          | Very important metric alongside with downloads                                                  |
| **`writer`**              | List of all the movie writers     | 🟠 Low                                          | None                                                                                           |


With this table, we can identify the important and the not so important variables of the Dataset, but before we can trasnform or delete this data, we will keep it to still analyse and describe it. For that, we weill begin by checking the Dataset dimensons with the **`.shape`** method:

In [82]:
# Checking dataset dimensions:
print(f'Total Variables: {data.shape[1]}\nTotal Rows: {data.shape[0]}')

Total Variables: 15
Total Rows: 20548


We can see that this is a rather large Dataset, containing more than 20k registry entries. Now, lets use the **`.info()`** method to gather more detailed information about these variables:

In [83]:
# Cheking general information about the Dataset:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20548 entries, 0 to 20547
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       20548 non-null  int64  
 1   IMDb-rating      19707 non-null  float64
 2   appropriate_for  11072 non-null  object 
 3   director         18610 non-null  object 
 4   downloads        20547 non-null  object 
 5   id               20548 non-null  int64  
 6   industry         20547 non-null  object 
 7   language         20006 non-null  object 
 8   posted_date      20547 non-null  object 
 9   release_date     20547 non-null  object 
 10  run_time         18780 non-null  object 
 11  storyline        18847 non-null  object 
 12  title            20547 non-null  object 
 13  views            20547 non-null  object 
 14  writer           18356 non-null  object 
dtypes: float64(1), int64(2), object(12)
memory usage: 2.4+ MB


Now we have a lot more information about the specific variables of the Dataset:

**Null Values**

- We have null values in all columns except the **`id`** variable.

**Data Types**

- In the Dataset, we have one **`float64`** variable, two **`int64`** variables and twelve **`object`** variables.

**Memory Usage**

- The Dataset, in it`s current form, consumes over 2.4 MB of memory.

**Proposed Changes**

- Make an acessement and deal with the null values;

- Change most of the Data Types to boost the memory efficiency. Also this is the most important change in this case because most of the variables are not in a apropriate data format, i.e. **`posted_date`** being a object data type instead of datetime and **`runtime`** being a object type instead of float or int.

Now, let`s check the exact amout of null values in each column, as well as the percentage that it represents:

In [84]:
# Checking total null values:
null_count = (data
              .isna()
              .sum()
              .sort_values(ascending=False))

null_count

appropriate_for    9476
writer             2192
director           1938
run_time           1768
storyline          1701
IMDb-rating         841
language            542
downloads             1
industry              1
posted_date           1
release_date          1
title                 1
views                 1
Unnamed: 0            0
id                    0
dtype: int64

In [85]:
# Check percentage of total size:
null_percentage = round((data.isna().sum()/data.shape[0]) * 100, 4)

null_percentage.sort_values(ascending=False)

appropriate_for    46.1164
writer             10.6677
director            9.4316
run_time            8.6042
storyline           8.2782
IMDb-rating         4.0929
language            2.6377
downloads           0.0049
industry            0.0049
posted_date         0.0049
release_date        0.0049
title               0.0049
views               0.0049
Unnamed: 0          0.0000
id                  0.0000
dtype: float64

immediately we can see that the variable **`appropriate_for`** is basically useless for the analysis. With 9.476 missing values representing more than 46% of the total data size, this variable is too empty to be filled in using different methods. For that reason, it will be dropped in the cleaning step.

The **`writer`** column also has a good amout of missing values, 2.192 nulls representing more than 10.5% of the total data size. As this variable was already deemed as having a low priority in Table 1.1 previously made, it will also be dropped in the cleaning phase.

The same cannot be said about the **`director`** variable, as altough it also has almost the same amount of null values as the **`writer`** column, it has a substancial potencial for analysis, being labeled as High priority in Table 1.1.

The **`runtime`** and **`storyline`** variables both have a similar amout of missing values (around 8.5% of the total data). The first of them can and will be filled in using the **`mode()`** of the variable (as this metric is not much affected by the presence of outliers and should maintain the data consistency), the second variable is going to be dropped as stated in Table 1.1.

All other variables have little to none Null values and can be easily filled in or dropped (like the **`Unnamed: 0`** column).


# **Data Prep, Cleaning and Feature Engineering**

---

Here we will make all of the changes that are needed in the dataset, all of them being described in the section above. For a detailed view of what exactly we will be doing here:

- Drop useless or undesirable variables;

- Rename the variables if necessary;

- Handle missing (null) values;

- Change data types;

- Handle duplicates, and

- Adding new features to the dataset (feature engineering).

Let's begin with the dropping of useless variables.

## **_Dropping Variables_**

As indicated by Table 1.1 and our subsequent analysis of null values, we have some variables that will be dropped and some that can be filled to use in the analysis. The following are the variables that will be deleted from the dataset:

**Variables to Drop**

- **`Unnamed: 0`**

- **`id`**

- **`storyline`**

- **`writer`**

- **`appropriate_for`**

For the manipulation of the dataset, we will create a copy of the Dataframe:

In [86]:
# Creating manipulation copy:
pirated_films = data.copy()

Now, we can drop the undesirable variables:

In [87]:
# Dropping variables:
pirated_films = (pirated_films
                 .drop(['Unnamed: 0', 'id', 'storyline', 'writer', 'appropriate_for'], 
                       axis=1))

In [88]:
# Cheking new data:
pirated_films.head()

IMDb-rating      director downloads             industry       language  \
0          4.8     John Swab       304  Hollywood / English        English   
1          6.4   Paul Ziller        73  Hollywood / English        English   
2          5.2  Ben Wheatley     1,427  Hollywood / English  English,Hindi   
3          8.1  Venky Atluri     1,549            Tollywood          Hindi   
4          4.6  Shaji Kailas       657            Tollywood          Hindi   

    posted_date release_date  run_time  \
0  20 Feb, 2023  Jan 28 2023       105   
1  20 Feb, 2023  Feb 05 2023        84   
2  20 Apr, 2021  Jun 18 2021  1h 47min   
3  20 Feb, 2023  Feb 17 2023       139   
4  20 Feb, 2023  Jan 26 2023       122   

                                        title   views  
0                                Little Dixie   2,794  
1  Grilling Season: A Curious Caterer Mystery   1,002  
2                                In the Earth  14,419  
3                                      Vaathi   4,878  
4                                       Alone   2,438

## **_Renaming Variables_**

With only the variables that we want in the dataset, we will change some variable names for the pourpose of standardizing the dataset and to have the best naming convention possible (short and descriptive).

In [89]:
# Renaming variables:
pirated_films = pirated_films.rename(columns={'IMDb-rating': 'imdb_user_rating',
                                              'director': 'film_director',
                                              'language': 'available_langs',
                                              'posted_date': 'platform_post_date',
                                              'release_date': 'worldwide_release',
                                              'run_time': 'run_time_min',
                                              'title': 'movie_title',
                                              'views': 'total_views'})

# Cheking data:
pirated_films.head()

imdb_user_rating film_director downloads             industry  \
0               4.8     John Swab       304  Hollywood / English   
1               6.4   Paul Ziller        73  Hollywood / English   
2               5.2  Ben Wheatley     1,427  Hollywood / English   
3               8.1  Venky Atluri     1,549            Tollywood   
4               4.6  Shaji Kailas       657            Tollywood   

  available_langs platform_post_date worldwide_release run_time_min  \
0         English       20 Feb, 2023       Jan 28 2023          105   
1         English       20 Feb, 2023       Feb 05 2023           84   
2   English,Hindi       20 Apr, 2021       Jun 18 2021     1h 47min   
3           Hindi       20 Feb, 2023       Feb 17 2023          139   
4           Hindi       20 Feb, 2023       Jan 26 2023          122   

                                  movie_title total_views  
0                                Little Dixie       2,794  
1  Grilling Season: A Curious Caterer Mystery       1,002  
2                                In the Earth      14,419  
3                                      Vaathi       4,878  
4                                       Alone       2,438

## **_Handling Null Values_**

Before we can determine the best practices for dealing with null values, let's check the null values of the variables still present in the dataset.

In [90]:
# Checking null values:
round((pirated_films.isna().sum()/pirated_films.shape[0]) * 100, 4)

imdb_user_rating      4.0929
film_director         9.4316
downloads             0.0049
industry              0.0049
available_langs       2.6377
platform_post_date    0.0049
worldwide_release     0.0049
run_time_min          8.6042
movie_title           0.0049
total_views           0.0049
dtype: float64

With that information, let's deal with these values, one variable at a time.

### **`imdb_user_rating`**

Let's get a better look at this variable:

In [91]:
pirated_films['imdb_user_rating'].describe()

count    19707.000000
mean         5.762151
std          1.374041
min          1.100000
25%          4.800000
50%          5.700000
75%          6.600000
max          9.900000
Name: imdb_user_rating, dtype: float64

For the fill method, we will use the **mean** of the total values of the variable:

In [92]:
# Defining mean:
mean_ratings = pirated_films['imdb_user_rating'].mean().round(1)

mean_ratings

5.8

In [93]:
# Placing mean in all null values at the variable:
pirated_films['imdb_user_rating'] = (pirated_films['imdb_user_rating']
                                     .fillna(mean_ratings))

pirated_films['imdb_user_rating'].isna().sum()

0

In [94]:
# Checking statistics:
pirated_films['imdb_user_rating'].describe()

count    20548.000000
mean         5.763700
std          1.345648
min          1.100000
25%          4.900000
50%          5.800000
75%          6.600000
max          9.900000
Name: imdb_user_rating, dtype: float64

### **`film_director`**

For this variable the procedure will be diferent, that's because it is a categorical (qualitative) variable. For these types of variables, the ideal fill method would be to use the mode of the column, the most recurring value.

But this column refers to the film director, in this case, if we assign all the null films to the same director (mode) the data will not be in a good shape. So, the null values will receive a string informing that the director was not informed.

In [95]:
# Filling nulls with custom string:
pirated_films['film_director'] = pirated_films['film_director'].fillna('Not Assigned')

pirated_films['film_director'].isna().sum()

0

In [96]:
# Checking variable:
pirated_films['film_director'].describe()

count            20548
unique            9673
top       Not Assigned
freq              1938
Name: film_director, dtype: object

Here we can already make out a characteristic of the data, in that the majority of films present in the dataset have no Director assigned to them.

### **`dowloads` and all other 0.0049% null variables**

Lets' check exactly how many null values are present in this variable, because, as the total null values check we did at the begining of this section shows, it has only 0.0049% of null values, along with other variables with the same amaout of nulls. Let's have a better look at this:

In [97]:
# Checking total values:
pirated_films['downloads'].isna().sum()

1

Apparently only one entry is missing the value, so could it be that all other variables that were shown to have only 0.0049% of nulls are in fact the same resigtry? Let's test this by querying the exact entry that has the null value on the `downloads` variable and see if the other ones are empty too:

In [98]:
# Checking the null entry:
pirated_films.query('downloads.isna()')

imdb_user_rating film_director downloads industry available_langs  \
149               7.1  Not Assigned       NaN      NaN             NaN   

    platform_post_date worldwide_release run_time_min movie_title total_views  
149                NaN               NaN          NaN         NaN         NaN

As we suspected, all of the variables that had 0.0049% of null values are in fact only one entry. As it doesn't even have the movie name, this entry is completely useless for the analysis and it will be dropped.

In [99]:
# Grabbing the index of the null registry:
drop_index = pirated_films.loc[(pirated_films['downloads'].isna())].index

# Dropping the entry:
pirated_films = pirated_films.drop(drop_index)

In [100]:
# Checking the null entry:
pirated_films.query('downloads.isna()')

Empty DataFrame
Columns: [imdb_user_rating, film_director, downloads, industry, available_langs, platform_post_date, worldwide_release, run_time_min, movie_title, total_views]
Index: []

We can also check if all the other variables that had 0.0049% of null values are now at 0%:

In [101]:
# Checking total nulls:
round((pirated_films.isna().sum()/pirated_films.shape[0]) * 100, 4)

imdb_user_rating      0.0000
film_director         0.0000
downloads             0.0000
industry              0.0000
available_langs       2.6330
platform_post_date    0.0000
worldwide_release     0.0000
run_time_min          8.5998
movie_title           0.0000
total_views           0.0000
dtype: float64

Indeed all other 0.0049% null variables are gone.

### **`available_langs`**

Another variable with null values refers to the available languages of the movie. As it is a categorical (qualitative) variable, we will use the mode as the fill method, as in, the most reocurring languages. But first, let's check the variable:

In [102]:
# Describing variable:
pirated_films['available_langs'].describe()

count       20006
unique       1168
top       English
freq        12657
Name: available_langs, dtype: object

As the describe method points to, the most reocurring language in the dataset in English, so this will be the value that will be placed in place of the null entries.

In [103]:
# Defining mode:
mode_fill = pirated_films['available_langs'].mode()[0] # English

# Filling the nulls:
pirated_films['available_langs'] = pirated_films['available_langs'].fillna(mode_fill)

In [104]:
# Checking nulls:
round((pirated_films.isna().sum()/pirated_films.shape[0]) * 100, 4)

imdb_user_rating      0.0000
film_director         0.0000
downloads             0.0000
industry              0.0000
available_langs       0.0000
platform_post_date    0.0000
worldwide_release     0.0000
run_time_min          8.5998
movie_title           0.0000
total_views           0.0000
dtype: float64

### **`run_time_min`**

The last variable with null values, let's get a closer look at it starting with the describe method:

In [105]:
# Describing variable:
pirated_films['run_time_min'].describe()

count     18780
unique      415
top          93
freq        652
Name: run_time_min, dtype: object

This variable will require more work. The current data type is object, but it must be numbers before we can fill the nulls, so, to start let's clean the data and change the data type:

In [106]:
# Checking values:
pirated_films['run_time_min'].sample(30).unique()

array(['125', '102', '100', '110', '143', '1h 33min', nan, '83', '197',
       '1h 27min', '84', '78', '92', '87', '1h 23min', '97', '104',
       '105 min', '109', '124', '96', '101', '1h 37min', '150', '142',
       '139'], dtype=object)

With us checking a random sample of the data, we can see that it is not in a appropriate format, as the variable name said it should be in minutes, we still have a lot of entries defining hours and, of course, the null values.

To clean this variable, a more robust function to do string manipulation is necessary:

In [107]:
# Creating function:
def convert_runtime(value):
    '''
    Converts a movie runtime string into only minutes.

    Args
    -----
        value (str): The movie runtime string in different formats of hours and minutes.

    Returns
    --------
        int: The movie runtime in minutes. If the input value is not a string or
            does not match the expected format, the original value is returned.

    Example
    --------
        >>> convert_runtime("1h 47min")
        107
        >>> convert_runtime("120 min")
        120
        >>> convert_runtime("2h 5m")
        125
        >>> convert_runtime("3h 25 min")
        205
    '''
    if not isinstance(value, str):
        return value
    
    # Pattern 1: xxh/hours xxm/min
    match = re.search(r'(\d+)\s*h(?:h|ours)?\s*(\d+)\s*(?:m|min)?', value)
    if match:
        hours, minutes = map(int, match.groups())
        return hours * 60 + minutes

    # Pattern 2: xxh/hr xxm/min
    match = re.search(r'(\d+)\s*(?:h|hr)\s*(\d+)\s*(?:m|min)?', value)
    if match:
        hours, minutes = map(int, match.groups())
        return hours * 60 + minutes

    # Pattern 3: xxh/hr xxm
    match = re.search(r'(\d+)\s*(?:h|hr)\s*(\d+)\s*m', value)
    if match:
        hours, minutes = map(int, match.groups())
        return hours * 60 + minutes

    # Pattern 4: xxm/min
    match = re.search(r'(\d+)\s*(?:m|min)', value)
    if match:
        return int(match.group(1))

    # Pattern 5: xxh/hours
    match = re.search(r'(\d+)\s*(?:h|hours)', value)
    if match:
        return int(match.group(1))

    return value


In [108]:
# Applying the function:
pirated_films['run_time_min'] = pirated_films['run_time_min'].apply(convert_runtime)

Now let's do a quick random check on the variable values:

In [109]:
# Checking results:
pirated_films['run_time_min'].sample(10)

13094     97
3949      92
12888    NaN
19545     98
10646    105
20121    170
2515     107
1725      87
18225    118
14357    134
Name: run_time_min, dtype: object

We can see that the data has been successfully transformed, where we now have all the run times in minutes only.
Now, we will fill the null values temporarily with 0 as a string, so we can chance the column data type to int. After that, we will fill the nulls (now the zeros) with the median of the values.

In [110]:
# Filling with zeros temporarily:
pirated_films['run_time_min'] = pirated_films['run_time_min'].fillna('0')

In [111]:
# Changing data type:
pirated_films['run_time_min'] = pirated_films['run_time_min'].astype('int')

Now that the variable is in the right data type, let's fill the nulls (zero values in this case) with the median of the values. The median is a good option here because these data are very likely to have outliers and to be skeewd, and as the median isn't very influenced by these factors, it becomes a good filling method. 

In [112]:
# Calculating the median value of the variable (ignoring the zeros):
median = np.median(pirated_films['run_time_min'][pirated_films['run_time_min'] != 0])

# Replacing the zeros with the median value:
pirated_films['run_time_min'][pirated_films['run_time_min'] == 0] = median

Finally, we can check the null quantity of our final dataset:

In [113]:
round((pirated_films.isna().sum()/pirated_films.shape[0]) * 100, 4)

imdb_user_rating      0.0
film_director         0.0
downloads             0.0
industry              0.0
available_langs       0.0
platform_post_date    0.0
worldwide_release     0.0
run_time_min          0.0
movie_title           0.0
total_views           0.0
dtype: float64

## **_Changing Data Types_**

With our dataset clean from null values and with only the variables we want still present, the next step is to adress the data types of the variables, to make sure they are in the best possible way for the analysis. Let's begin by checking our dtypes:

In [114]:
# Checking data types:
pirated_films.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20547 entries, 0 to 20547
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   imdb_user_rating    20547 non-null  float64
 1   film_director       20547 non-null  object 
 2   downloads           20547 non-null  object 
 3   industry            20547 non-null  object 
 4   available_langs     20547 non-null  object 
 5   platform_post_date  20547 non-null  object 
 6   worldwide_release   20547 non-null  object 
 7   run_time_min        20547 non-null  int32  
 8   movie_title         20547 non-null  object 
 9   total_views         20547 non-null  object 
dtypes: float64(1), int32(1), object(8)
memory usage: 1.6+ MB


Things to note:

**Memory Consumption**

- With the current data types, the dataset uses up to 1.6 MB of memory in the system. With the changes we are going to make, this consumption should drop.

**Good Data Types**

We have some variables that don't require any change:

- `imdb_user_rating` in **float64**

- `film_director` in **object**

- `industry` in **object**

- `available_langs` in **object**

- `run_time_min` in **int32**

- `movie_title` in **object**

**Wrong Data Types**

For variables that need changes we have:

- `downloads` from **object** to **int**

- `platform_post_date` from **object** to **datetime**

- `worldwide_release` from **object** to **datetime**

- `total_views` from **object** to **int**

First we will deal with the **datetime** columns alone, as they very straight forward. The other two variables (`downloads` and `total_views`) require a little bit more work.

### **Altering `platform_post_date` and `worldwide_release`**

In [115]:
# Changing Data Types:
pirated_films['platform_post_date'] = pd.to_datetime(pirated_films['platform_post_date'])

pirated_films['worldwide_release'] = pd.to_datetime(pirated_films['worldwide_release'])

In [116]:
# Checking new data types:
pirated_films.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20547 entries, 0 to 20547
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   imdb_user_rating    20547 non-null  float64       
 1   film_director       20547 non-null  object        
 2   downloads           20547 non-null  object        
 3   industry            20547 non-null  object        
 4   available_langs     20547 non-null  object        
 5   platform_post_date  20547 non-null  datetime64[ns]
 6   worldwide_release   20547 non-null  datetime64[ns]
 7   run_time_min        20547 non-null  int32         
 8   movie_title         20547 non-null  object        
 9   total_views         20547 non-null  object        
dtypes: datetime64[ns](2), float64(1), int32(1), object(6)
memory usage: 1.6+ MB


In [117]:
# Checking format:
pirated_films.head()

imdb_user_rating film_director downloads             industry  \
0               4.8     John Swab       304  Hollywood / English   
1               6.4   Paul Ziller        73  Hollywood / English   
2               5.2  Ben Wheatley     1,427  Hollywood / English   
3               8.1  Venky Atluri     1,549            Tollywood   
4               4.6  Shaji Kailas       657            Tollywood   

  available_langs platform_post_date worldwide_release  run_time_min  \
0         English         2023-02-20        2023-01-28           105   
1         English         2023-02-20        2023-02-05            84   
2   English,Hindi         2021-04-20        2021-06-18           107   
3           Hindi         2023-02-20        2023-02-17           139   
4           Hindi         2023-02-20        2023-01-26           122   

                                  movie_title total_views  
0                                Little Dixie       2,794  
1  Grilling Season: A Curious Caterer Mystery       1,002  
2                                In the Earth      14,419  
3                                      Vaathi       4,878  
4                                       Alone       2,438

### **Altering `downloads` and `total_views`**

Before we can change the data type of this variable, we need to clean it from any special characters like commas and dots. Let's do a quick check on them:

In [118]:
# Checking values:
pirated_films['downloads'].head(15)

0        304
1         73
2      1,427
3      1,549
4        657
5        746
6      5,332
7      1,781
8        458
9      1,965
10       742
11    12,954
12     2,253
13    14,867
14       463
Name: downloads, dtype: object

In [119]:
# Checking more values from the other variable:
pirated_films['total_views'].head(15)

0      2,794
1      1,002
2     14,419
3      4,878
4      2,438
5      2,940
6     11,978
7     18,225
8      6,912
9      9,710
10     4,618
11    35,831
12     5,468
13    39,399
14     5,763
Name: total_views, dtype: object

We can see that yes, both variables have special characters and require cleaning before we can change the data types. Let's create a general cleaning function that can be applied to both variables to clean them.

In [120]:
# Creating cleaning function:
def number_extractor(data: pd.DataFrame):
    '''
    Applies a regular expression to remove all non-numeric characters from the 'total_views' and 'downloads'
    columns of a given pandas DataFrame.

    Parameters:
    -----------
    data : pandas DataFrame
        The input DataFrame to process. The DataFrame must contain the 'total_views' and 'downloads' columns.

    Returns:
    --------
    pandas DataFrame:
        A copy of the input DataFrame where the 'total_views' and 'downloads' columns have been modified to
        only contain numeric characters.
    '''
    chars_to_remove = r'[^0-9]+'

    data['total_views'] = data['total_views'].apply(
        lambda clean_values:
            re.sub(chars_to_remove, '', str(clean_values))
    )

    data['downloads'] = data['downloads'].apply(
        lambda clean_values:
            re.sub(chars_to_remove, '', str(clean_values))
    )

    return data

In [121]:
# Applying the function:
pirated_films = number_extractor(pirated_films)

# Checking results:
pirated_films['total_views'].head()

0     2794
1     1002
2    14419
3     4878
4     2438
Name: total_views, dtype: object

In [122]:
# Checking results:
pirated_films['downloads'].head()

0     304
1      73
2    1427
3    1549
4     657
Name: downloads, dtype: object

With the variables clean, we can pass them to int:

In [123]:
# Changing data types:
pirated_films = pirated_films.astype({'downloads': 'int',
                                      'total_views': 'int'})

In [124]:
# Checking new data:
pirated_films.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 20547 entries, 0 to 20547
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   imdb_user_rating    20547 non-null  float64       
 1   film_director       20547 non-null  object        
 2   downloads           20547 non-null  int32         
 3   industry            20547 non-null  object        
 4   available_langs     20547 non-null  object        
 5   platform_post_date  20547 non-null  datetime64[ns]
 6   worldwide_release   20547 non-null  datetime64[ns]
 7   run_time_min        20547 non-null  int32         
 8   movie_title         20547 non-null  object        
 9   total_views         20547 non-null  int32         
dtypes: datetime64[ns](2), float64(1), int32(3), object(4)
memory usage: 1.5+ MB


With all corrections needed made to the dataset, we can check for duplicates in the data:

In [125]:
# Checking duplicates:
pirated_films[pirated_films.duplicated()]

Empty DataFrame
Columns: [imdb_user_rating, film_director, downloads, industry, available_langs, platform_post_date, worldwide_release, run_time_min, movie_title, total_views]
Index: []

We see that we don't have any duplicate values in our data. With that, we can move to our next step: Adding new features to our dataset!

## **_Feature Engineering_**

To define what additional features our dataset can have to boost our analysis capabilities, let's take a look at it's variables:

In [126]:
# Checking dataset:
pirated_films.head()

imdb_user_rating film_director  downloads             industry  \
0               4.8     John Swab        304  Hollywood / English   
1               6.4   Paul Ziller         73  Hollywood / English   
2               5.2  Ben Wheatley       1427  Hollywood / English   
3               8.1  Venky Atluri       1549            Tollywood   
4               4.6  Shaji Kailas        657            Tollywood   

  available_langs platform_post_date worldwide_release  run_time_min  \
0         English         2023-02-20        2023-01-28           105   
1         English         2023-02-20        2023-02-05            84   
2   English,Hindi         2021-04-20        2021-06-18           107   
3           Hindi         2023-02-20        2023-02-17           139   
4           Hindi         2023-02-20        2023-01-26           122   

                                  movie_title  total_views  
0                                Little Dixie         2794  
1  Grilling Season: A Curious Caterer Mystery         1002  
2                                In the Earth        14419  
3                                      Vaathi         4878  
4                                       Alone         2438

Some additional features that we can add to the dataset are:

- `days_to_piracy`: A quantitative variable that contains the difference in days between the film release date and the date it was posted on the piracy site.

- `avg_views_per_download`: The total number of views divided by the total number of downloads, indicating how many people viewed the movie for each download.

- `avg_downloads_per_day`: A new feature that calculates the average number of downloads per day since the movie was posted on the piracy website. This could help us understand the popularity of the movie among pirates over time.

### **Creating Variable: `days_to_piracy`**

This variable will lead to intresting possibilities of analysis, allowing us to understand better the timing between a movie release and it's pirated version being available for download. This variable will also play a role in creating the `popularity_score` variable later.

To create this variable, we will subtract the day of the worldwide release and the day that the movie got launched in the piracy site.

In [127]:
# Creating new variable to store the difference in days.
# This variable is temporary, it will be used to create the oficial variable:
pirated_films['days_temp'] = pirated_films['platform_post_date'] - pirated_films['worldwide_release']

In [128]:
# Checking variable:
pirated_films['days_temp'].head()

0    23 days
1    15 days
2   -59 days
3     3 days
4    25 days
Name: days_temp, dtype: timedelta64[ns]

Now that we have this variable, we will use it to grab only the number of days:

In [129]:
# Creating the oficial variable and deleting the temporary one:
pirated_films['days_to_piracy'] = pirated_films['days_temp'].dt.days

# Dropping temporary variable:
pirated_films = pirated_films.drop('days_temp', axis=1)

In [130]:
# Checking results:
pirated_films.head()

imdb_user_rating film_director  downloads             industry  \
0               4.8     John Swab        304  Hollywood / English   
1               6.4   Paul Ziller         73  Hollywood / English   
2               5.2  Ben Wheatley       1427  Hollywood / English   
3               8.1  Venky Atluri       1549            Tollywood   
4               4.6  Shaji Kailas        657            Tollywood   

  available_langs platform_post_date worldwide_release  run_time_min  \
0         English         2023-02-20        2023-01-28           105   
1         English         2023-02-20        2023-02-05            84   
2   English,Hindi         2021-04-20        2021-06-18           107   
3           Hindi         2023-02-20        2023-02-17           139   
4           Hindi         2023-02-20        2023-01-26           122   

                                  movie_title  total_views  days_to_piracy  
0                                Little Dixie         2794              23  
1  Grilling Season: A Curious Caterer Mystery         1002              15  
2                                In the Earth        14419             -59  
3                                      Vaathi         4878               3  
4                                       Alone         2438              25

Just by looking at the first values of the new variable we can see that it gives very helpfull information abaout the films, such as the third entry in the dataset that says that the movie was already pirated and available in the piracy web site 59 days before even launching worldwide.

Let's check the data type of the variable:

In [131]:
# Checking data type:
pirated_films['days_to_piracy'].dtype

dtype('int64')

The variable came in the int64 data type. To give it the best data type possible, let's see if we can change it to int32:

In [132]:
# Checking min and max values of the new variable:
print(f'Max value: {pirated_films["days_to_piracy"].max()}')
print(f'Min value: {pirated_films["days_to_piracy"].min()}')

Max value: 32863
Min value: -4298


We almost can change it to int16! The limit of int16 is 32767 and in the data we have values up to 32863, close, but enough to be necessary the implementation of int32.

In [133]:
# Changing data type to int32:
pirated_films = pirated_films.astype({'days_to_piracy': 'int32'})

In [134]:
# Checking final data types:
pirated_films['days_to_piracy'].dtype

dtype('int32')

### **Creating Variable: `avg_views_per_download`**

This variable, togheter with the next one, will make it possible for us to understand the relevance and popularity of each movie pirated on the website, more specicaly this variable allow us to beter understand how frequently pirates watch the movies they download.

In [135]:
# Creating new variable:
pirated_films['avg_views_per_download'] = pirated_films['total_views'] / pirated_films['downloads']

# Visualizing data:
pirated_films.head()

imdb_user_rating film_director  downloads             industry  \
0               4.8     John Swab        304  Hollywood / English   
1               6.4   Paul Ziller         73  Hollywood / English   
2               5.2  Ben Wheatley       1427  Hollywood / English   
3               8.1  Venky Atluri       1549            Tollywood   
4               4.6  Shaji Kailas        657            Tollywood   

  available_langs platform_post_date worldwide_release  run_time_min  \
0         English         2023-02-20        2023-01-28           105   
1         English         2023-02-20        2023-02-05            84   
2   English,Hindi         2021-04-20        2021-06-18           107   
3           Hindi         2023-02-20        2023-02-17           139   
4           Hindi         2023-02-20        2023-01-26           122   

                                  movie_title  total_views  days_to_piracy  \
0                                Little Dixie         2794              23   
1  Grilling Season: A Curious Caterer Mystery         1002              15   
2                                In the Earth        14419             -59   
3                                      Vaathi         4878               3   
4                                       Alone         2438              25   

   avg_views_per_download  
0                9.190789  
1               13.726027  
2               10.104415  
3                3.149128  
4                3.710807

As we will use this variable to further analysis, we won't round the values, instead we will keep them with the full precision of the values.

### **Creating Variable: `avg_downloads_per_day`**

In junction with the `avg_views_per_download` variable, this column will allow us to beter understand the popularity and preference of movies in the piracy platform.

One thing we must remember is that we have movies that got pirated the same day it was released worldwide. That means that our calculation could give a "inf" value, since we could be dividing by 0 (number of days), to avoid tis, let's fill the inf values with the number of dowloads.

In [136]:
# Define function:
def add_downloads_per_day(df):
    # calculate downloads per day
    downloads_per_day = df['downloads'] / df['days_to_piracy']

    # Checks if days_to_piracy is equal to 0 and replaces the corresponding values in 
    # downloads_per_day with the downloads values to avoid division by zero.
    downloads_per_day[df['days_to_piracy'] == 0] = df['downloads'][df['days_to_piracy'] == 0]

    # add new column to DataFrame
    df['avg_downloads_per_day'] = np.abs(downloads_per_day)

    return df

In [137]:
# Aplying the function:
pirated_films = add_downloads_per_day(pirated_films)

# Checking values:
pirated_films.head()

imdb_user_rating film_director  downloads             industry  \
0               4.8     John Swab        304  Hollywood / English   
1               6.4   Paul Ziller         73  Hollywood / English   
2               5.2  Ben Wheatley       1427  Hollywood / English   
3               8.1  Venky Atluri       1549            Tollywood   
4               4.6  Shaji Kailas        657            Tollywood   

  available_langs platform_post_date worldwide_release  run_time_min  \
0         English         2023-02-20        2023-01-28           105   
1         English         2023-02-20        2023-02-05            84   
2   English,Hindi         2021-04-20        2021-06-18           107   
3           Hindi         2023-02-20        2023-02-17           139   
4           Hindi         2023-02-20        2023-01-26           122   

                                  movie_title  total_views  days_to_piracy  \
0                                Little Dixie         2794              23   
1  Grilling Season: A Curious Caterer Mystery         1002              15   
2                                In the Earth        14419             -59   
3                                      Vaathi         4878               3   
4                                       Alone         2438              25   

   avg_views_per_download  avg_downloads_per_day  
0                9.190789              13.217391  
1               13.726027               4.866667  
2               10.104415              24.186441  
3                3.149128             516.333333  
4                3.710807              26.280000

In [138]:
# Checking datatype:
pirated_films['avg_downloads_per_day'].dtype

dtype('float64')

## **_Rearranging Variables_**

With our variables cleaned, sorted and with our new variables created, we can rearrange the sequence of the columns to get a better order of visualization.

In [139]:
# Rearranging variables:
new_order = ['movie_title', 'film_director', 'industry', 'available_langs', 
             'run_time_min', 'imdb_user_rating', 'worldwide_release', 
             'platform_post_date', 'days_to_piracy', 'avg_views_per_download', 
             'avg_downloads_per_day', 'downloads', 'total_views']

# Defining the order:
pirated_films = pirated_films.reindex(columns=new_order)

# Checking final dataset:
pirated_films.head(10)

movie_title       film_director  \
0                                Little Dixie           John Swab   
1  Grilling Season: A Curious Caterer Mystery         Paul Ziller   
2                                In the Earth        Ben Wheatley   
3                                      Vaathi        Venky Atluri   
4                                       Alone        Shaji Kailas   
5                           Anubhavinchu Raja  Srinivas Gavireddy   
6                     WWE Elimination Chamber        Not Assigned   
7                                     Sharper      Benjamin Caron   
8                                Four Samosas         Ravi Kapoor   
9                               Blood Harvest        Danny LeGare   

              industry available_langs  run_time_min  imdb_user_rating  \
0  Hollywood / English         English           105               4.8   
1  Hollywood / English         English            84               6.4   
2  Hollywood / English   English,Hindi           107               5.2   
3            Tollywood           Hindi           139               8.1   
4            Tollywood           Hindi           122               4.6   
5            Tollywood           Hindi           131               5.4   
6            Wrestling         English           200               5.8   
7  Hollywood / English         English           116               6.5   
8  Hollywood / English         English            80               6.9   
9  Hollywood / English         English            80               4.2   

  worldwide_release platform_post_date  days_to_piracy  \
0        2023-01-28         2023-02-20              23   
1        2023-02-05         2023-02-20              15   
2        2021-06-18         2021-04-20             -59   
3        2023-02-17         2023-02-20               3   
4        2023-01-26         2023-02-20              25   
5        2021-11-26         2023-02-20             451   
6        2023-02-18         2023-02-19               1   
7        2023-02-17         2023-02-13              -4   
8        2022-12-02         2023-02-18              78   
9        2023-02-07         2023-02-18              11   

   avg_views_per_download  avg_downloads_per_day  downloads  total_views  
0                9.190789              13.217391        304         2794  
1               13.726027               4.866667         73         1002  
2               10.104415              24.186441       1427        14419  
3                3.149128             516.333333       1549         4878  
4                3.710807              26.280000        657         2438  
5                3.941019               1.654102        746         2940  
6                2.246437            5332.000000       5332        11978  
7               10.233015             445.250000       1781        18225  
8               15.091703               5.871795        458         6912  
9                4.941476             178.636364       1965         9710

With our dataset ready for analysis, we will create a new Notebook that will hold the Univariate Analysis for the project. This will be done for a couple of reasons:

- If we were to do the entire EDA on a single notebook, it would end up to large, that would make it harder and more tedious to navigate the document and find the information you want.

- With multiple notebooks for each step of the project as large as this one, it makes very simple and easy to understand where certain parts of the EDA are. Want to see the cleaning phase? this is the notebook, want to see the plots and analysis? that is the notebook, etc.

So to continue on to the next notebook, let's save our dataset as a new CSV.

In [140]:
# Creating a new clean CSV:
pirated_films.to_csv('data/pirated_films.csv', index=False, sep=',')